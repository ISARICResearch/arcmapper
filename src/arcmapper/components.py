from dash import html, dcc
import dash_bootstrap_components as dbc


def select(id: str, values: list[str], default: str | None = None) -> dbc.Select:
    return dbc.Select(
        id=id,
        value=default if default else values[0],
        options=[{"label": v, "value": v} for v in values],
    )


upload_form = dbc.Container(
    html.Div(
        dbc.Form(
            [
                dbc.Row(
                    [
                        html.P(
                            [
                                html.Strong("SOURCE: "),
                                "ARCmapper supports data dictionaries in CSV, XLSX, JSON schema, or you can upload sample data.",
                            ]
                        )
                    ]
                ),
                dbc.Row(
                    [
                        dbc.Label("Responses column", width="auto"),
                        dbc.Col(
                            dbc.Input(
                                id="upload-col-responses",
                                type="text",
                                value="Choices, Calculations, OR Slider Labels",
                            ),
                            className="me-3",
                        ),
                        dbc.Label("Description column", width="auto"),
                        dbc.Col(
                            dbc.Input(
                                id="upload-col-description",
                                type="text",
                                placeholder="Defaults to longest column",
                                value="Field Label",
                            ),
                            className="me-3",
                        ),
                        dbc.Col(
                            dcc.Upload(
                                id="upload-input-file",
                                children=html.Div(
                                    "Upload data dictionary",
                                    style={
                                        "background": "#316cf4",
                                        "color": "white",
                                        "padding": "0.4em",
                                        "borderRadius": "5px",
                                        "cursor": "pointer",
                                    },
                                ),
                            ),
                            className="me-3",
                        ),
                        dcc.Store(id="upload-data-dictionary"),
                    ],
                    className="g-2",
                ),
                dbc.Row(id="upload-status"),
            ]
        ),
        style={
            "border": "1px solid silver",
            "border-radius": "0.4em",
            "padding": "1em",
        },
    ),
    style={"margin-top": "1em"},
)

arc_form = dbc.Container(
    html.Div(
        dbc.Form(
            [
                dbc.Row(
                    html.P(
                        [
                            html.Strong("TARGET: "),
                            "Choose the target ARC version and select method and method parameters",
                        ]
                    )
                ),
                dbc.Row(
                    [
                        dbc.Label("Target ARC version", width="auto"),
                        dbc.Col(
                            select("arc-version", ["1.0.0", "1.0.1"]),
                            className="me-3",
                        ),
                        dbc.Label("Mapping method", width="auto"),
                        dbc.Col(
                            dbc.Select(
                                id="arc-mapping-method",
                                options=[
                                    {"label": "TF-IDF", "value": "tf-idf"},
                                    {
                                        "label": "Sentence Transformers",
                                        "value": "sbert",
                                    },
                                ],
                                value="tf-idf",
                            ),
                            className="me-3",
                        ),
                        dbc.Label("Number of matches", width="auto"),
                        dbc.Col(
                            dbc.Input(
                                id="arc-num-matches",
                                type="number",
                                min=2,
                                max=10,
                                step=1,
                                value=3,
                            ),
                        ),
                        dbc.Label("Threshold", width="auto"),
                        dbc.Col(
                            dbc.Input(
                                id="arc-threshold",
                                type="number",
                                min=0.1,
                                max=1,
                                step=0.1,
                                value=0.3,
                            ),
                        ),
                        dbc.Col(dbc.Button("Map to ARC", id="map-btn"), width="auto"),
                    ],
                    className="g-2",
                ),
            ]
        ),
        style={
            "border": "1px solid silver",
            "border-radius": "0.4em",
            "padding": "1em",
        },
    ),
    style={"margin-top": "1em"},
)
