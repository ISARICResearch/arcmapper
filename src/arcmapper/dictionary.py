"""Types and methods for loading data dictionaries"""

import json
import operator
from typing import Any, NamedTuple

import pandas as pd
from pandas.api.types import is_object_dtype
from .types import DataType
from .util import read_data, parse_redcap_response

RESPONSE_PARSERS = {"redcap": parse_redcap_response}


class DictionaryField(NamedTuple):
    "Data dictionary field"

    variable: str
    description: str
    responses: str | None
    type: DataType


def read_data_dictionary(
    source: str | pd.DataFrame,
    variable_field: str | None = None,
    description_field: str | None = None,
    type_field: str | None = None,
    response_field: str | None = None,
    response_func: str | None = None,
) -> pd.DataFrame:
    """Reads from data dictionary file or data frame

    By default, form responses (categorical enumerations) are not read, as
    there is no standard format. This can be changed by setting *both*
    response_field and response_func to return responses.

    Parameters
    ----------
    source
        Data dictionary to read. Can also pass a dataframe instead of a file
    variable_field
        Field to use for variable name. If not specified, uses the first
        column
    description_field
        Field to use for description. If not specified, is taken to be the
        stringly typed column with the maximum mean length.
    type_field
        Field to use for type information. If not specified, every type
        defaults to 'string'
    response_field
        Response field to use
    response_func
        Function that takes a string and returns a Responses type

    Returns
    -------
    pd.DataFrame
        Data dictionary
    """
    dd = read_data(source)
    variable_field = variable_field or dd.columns[0]
    if description_field is None:
        description_field = max(
            (
                (c, dd[c].map(lambda x: len(x) if isinstance(x, str) else 0).mean())
                for c in dd.columns
                if is_object_dtype(dd[c])
            ),
            key=operator.itemgetter(1),
        )[0]
    if (response_field is None) ^ (response_func is None):
        raise ValueError("Both response_field and response_func have to be specified")
    assert (
        response_func in RESPONSE_PARSERS
    ), f"Unknown response parser: {response_func}"
    return pd.DataFrame(
        [
            DictionaryField(
                row[variable_field],
                row[description_field],
                RESPONSE_PARSERS[response_func](row[response_field])
                if response_field and isinstance(row[response_field], str)
                else None,
                row[type_field] if type_field else "string",
            )
            for row in dd.to_dict(orient="records")
        ]
    )


def read_from_data(data: str | pd.DataFrame) -> pd.DataFrame:
    "Infers data dictionary from sample data"
    raise NotImplementedError


def read_from_jsonschema(data: str | dict[str, Any]) -> pd.DataFrame:  # type: ignore
    "Returns DataDictionary from JSON Schema file or data"
    if isinstance(data, str):
        data: dict[str, Any] = json.loads(data)
    dd = []
    variables = data["properties"]
    for v in variables:
        t = variables[v].get("type","string")
        if "enum" in variables[v]:
            dd.append((v, variables[v].get("description", ""), [(x, x) for x in variables[v]["enum"]], 'categorical'))
        else:
            dd.append((v, variables[v].get("description", ""), None, t))
    return pd.DataFrame(dd, columns=["variable", "description", "responses", "type"])
