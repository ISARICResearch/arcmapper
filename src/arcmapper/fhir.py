"""Final mapping of data dictionary to FHIR

Mapping file MUST conform to specification at
https://fhirflat.readthedocs.io/en/latest/spec/mapping.html
"""

import warnings
from pathlib import Path

import pandas as pd

from .strategies import infer_response_mapping

VALID_FHIR_RESOURCES = [
    "Condition",
    "DiagnosticReport",
    "Encounter",
    "Immunization",
    "MedicationAdministration",
    "MedicationStatement",
    "Observation",
    "Patient",
    "Procedure",
    "Specimen",
]


class FHIRMapping:
    "Loads mapping file from a Excel (XLSX) sheet"

    def __init__(self, file: str | Path):
        path = Path(file)
        if path.suffix != ".xlsx":
            raise ValueError("FHIRMapping only supports Excel sheets at the moment")
        index = pd.read_excel(path)
        if "Resources" not in index.columns:
            raise ValueError(
                "Required 'Resources' column not present in FHIR mapping file"
            )
        self.resources = sorted(set(index.Resources) & set(VALID_FHIR_RESOURCES))
        if "Patient" not in self.resources:
            raise ValueError(
                "Required FHIR mapping for FHIR resource 'Patient' not found in mapping file"
            )
        self.path = path

    def get_resource(self, resource: str) -> pd.DataFrame:
        "Gets resource from FHIR mapping Excel sheet"
        resource = resource.capitalize()  # capitalize first letter
        if resource not in self.resources:
            raise ValueError(
                f"Resource '{resource}' not found, valid resources: {self.resources}"
            )
        df = pd.read_excel(self.path, sheet_name=resource)

        # forward fill NaNs to enable merge with mapping frame
        df["raw_variable"] = df["raw_variable"].ffill()
        return df.rename(
            columns={"raw_variable": "arc_variable", "raw_response": "arc_response"}
        )


def merge(
    draft: pd.DataFrame, mapping: FHIRMapping, resources: list[str] = []
) -> dict[str, pd.DataFrame]:
    out = {}
    draft = infer_response_mapping(draft)
    for resource in resources or mapping.resources:
        # first generate choice responses for each mapping
        if resource not in mapping.resources:
            warnings.warn(
                f"Resource requested to be mapped but not found in mapping file: {resource}"
            )
        out[resource] = draft.merge(
            mapping.get_resource(resource), on=["arc_variable", "arc_response"]
        )
    return out


def format_merge(merged_data, selected_columns: list[str] | None = None):
    out = ""
    selected_columns = selected_columns or [
        "raw_variable",
        "raw_response",
        "arc_variable",
        "arc_response",
        "raw_description",
        "arc_description",
    ]
    for resource in merged_data:
        out += "{{{ resource " + resource + "\n"
        out += merged_data[resource][selected_columns].to_csv(index=False, sep="\t")
        out += "}}}\n"
    return out.strip()
