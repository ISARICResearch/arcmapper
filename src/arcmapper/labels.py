"Labels for buttons and forms"

MAP_TO_ARC = "→ Map to ARC"
UPLOAD_DATA_DICTIONARY = "↟ Upload data dictionary"
SAVE_INTERMEDIATE_FILE = "↓ Save intermediate file"
LOAD_INTERMEDIATE_FILE = "↑ Load intermediate file"
DOWNLOAD_FHIRFLAT_MAPPING = "▼ Download FHIRflat mapping"
