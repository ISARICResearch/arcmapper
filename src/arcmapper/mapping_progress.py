"""Mapping progress for ARC to FHIR.

ARC is the data dictionary containing the set of questions with possible
responses being developed by ISARIC at
https://github.com/ISARICResearch/DataPlatform/tree/main/ARCH

Each of the ARC variables is mapped to a FHIR resource with the corresponding
resource elements. FHIR resource descriptions can be obtained at
https://hl7.org/fhir/resourcelist.html, and the mapping file specification is
at https://fhirflat.readthedocs.io/en/latest/spec/mapping.html
"""


