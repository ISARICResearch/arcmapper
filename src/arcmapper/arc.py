"Module to read ARC schema"

import pandas as pd

from .types import DataType
from .util import read_csv_with_encoding_detection
from .dictionary import read_data_dictionary


def arc_schema_url(arc_version: str) -> str:
    return f"https://github.com/ISARICResearch/DataPlatform/raw/refs/heads/main/ARCH/ARCH{arc_version}/ARCH.csv"



def read_arc_schema(
    arc_version_or_file: str, preset: str | None = None
) -> pd.DataFrame:
    types_mapping: dict[str, DataType] = {
        "radio": "categorical",
        "number": "number",
        "text": "string",
        "date_dmy": "date",
        "checkbox": "categorical",
        "dropdown": "categorical",
        "datetime_dmy": "date",
    }
    arc_location = (
        arc_version_or_file
        if arc_version_or_file.endswith(".csv")
        else arc_schema_url(arc_version_or_file)
    )
    arc = read_csv_with_encoding_detection(arc_location)
    arc["Description"] = arc.Question + " " + arc.Definition
    arc["Type"] = arc.Type.map(types_mapping)
    dd = read_data_dictionary(
        arc,
        variable_field="Variable",
        description_field="Description",
        type_field="Type",
        response_field="Answer Options",
        response_func="redcap",
    )
    dd = dd[~pd.isna(dd.description)]
    if preset:
        preset_col = "preset_" + preset
        if preset_col not in arc.columns:
            raise ValueError(f"No such preset column exists in ARC: {preset_col}")
        dd = dd[dd[preset_col] == 1]
    return dd
