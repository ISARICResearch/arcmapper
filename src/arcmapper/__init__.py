import pandas as pd
from .arc import read_arc_schema
from .dictionary import read_data_dictionary
from .strategies import tf_idf, sbert


def map(
    method: str,
    dictionary: pd.DataFrame,
    arc: pd.DataFrame,
    num_matches: int = 3,
    **kwargs,
) -> pd.DataFrame:
    match method:
        case "tf-idf":
            return tf_idf(dictionary, arc, num_matches)
        case "sbert":
            return sbert(dictionary, arc, num_matches=num_matches)
        case _:
            raise ValueError(f"Unknown mapping method: {method}")


__all__ = ["read_arc_schema", "map", "read_data_dictionary"]
