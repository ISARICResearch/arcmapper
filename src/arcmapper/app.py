"""Dash frontend for the arcmapper library"""

import pandas as pd
import dash
from dash import dcc, html, ctx, callback, dash_table, Input, Output, State
import dash_bootstrap_components as dbc


from .components import select
from .util import read_upload_data
from .dictionary import read_data_dictionary
from .strategies import map as map_data_dictionary_to_arc
from .arc import read_arc_schema

app = dash.Dash("arcmapper", external_stylesheets=[dbc.themes.BOOTSTRAP])
app.title = "ARCMapper"

navbar = dbc.Navbar(
    dbc.Container(
        [
            html.A(
                # Use row and col to control vertical alignment of logo / brand
                dbc.Row(
                    [
                        dbc.Col(
                            html.Img(
                                src=dash.get_asset_url("ISARIC_logo_wh.png"),
                                height="60px",
                            )
                        ),
                        dbc.Col(dbc.NavbarBrand("ARCMapper", className="ms-2")),
                    ],
                    align="center",
                    className="g-0",
                ),
                href="https://isaric.org/",
                style={"textDecoration": "none"},
            ),
            dbc.NavbarToggler(id="navbar-toggler", n_clicks=0),
        ]
    ),
    color="#BA0225",
    dark=True,
)

upload_form = dbc.Container(
    html.Div(
        dbc.Form(
            [
                dbc.Row(
                    [
                        html.P(
                            [
                                html.Strong("SOURCE: "),
                                "ARCmapper supports data dictionaries in CSV, XLSX, JSON schema, or you can upload sample data.",
                            ]
                        )
                    ]
                ),
                dbc.Row(
                    dbc.Col(
                        dbc.Switch(
                            id="upload-is-sample-data",
                            label="Uploaded file is sample data, not a data dictionary. Data will be sent to server, only use on local deployments",
                            disabled=True,
                        )
                    )
                ),
                dbc.Row(
                    [
                        dbc.Label("Responses column", width="auto"),
                        dbc.Col(
                            dbc.Input(
                                id="upload-col-responses",
                                type="text",
                                value="Choices, Calculations, OR Slider Labels",
                            ),
                            className="me-3",
                        ),
                        dbc.Label("Description column", width="auto"),
                        dbc.Col(
                            dbc.Input(
                                id="upload-col-description",
                                type="text",
                                placeholder="Defaults to longest column",
                                value="Field Label",
                            ),
                            className="me-3",
                        ),
                        dbc.Col(
                            dcc.Upload(
                                id="upload-input-file",
                                children=html.Div(
                                    "Drag and drop or select file",
                                    style={
                                        "border": "1px dashed silver",
                                        "padding": "0.3em",
                                    },
                                ),
                            ),
                            className="me-3",
                        ),
                        dbc.Col(
                            dbc.Button(
                                "Upload", id="upload-btn", color="primary", n_clicks=0
                            )
                        ),
                        dcc.Store(id="upload-data-dictionary"),
                    ],
                    className="g-2",
                ),
                dbc.Row(id="upload-status"),
            ]
        ),
        style={
            "border": "1px solid silver",
            "border-radius": "0.4em",
            "padding": "1em",
        },
    ),
    style={"margin-top": "1em"},
)

arc_form = dbc.Container(
    html.Div(
        dbc.Form(
            [
                dbc.Row(
                    html.P(
                        [
                            html.Strong("TARGET: "),
                            "Choose the target ARC version and select method and method parameters",
                        ]
                    )
                ),
                dbc.Row(
                    [
                        dbc.Label("Target ARC version", width="auto"),
                        dbc.Col(
                            select("arc-version", ["1.0.0", "1.0.1"]),
                            className="me-3",
                        ),
                        dbc.Label("Mapping method", width="auto"),
                        dbc.Col(
                            dbc.Select(
                                id="arc-mapping-method",
                                options=[
                                    {"label": "TF-IDF", "value": "tf-idf"},
                                    {
                                        "label": "Sentence Transformers",
                                        "value": "sbert",
                                    },
                                ],
                                value="tf-idf",
                            ),
                            className="me-3",
                        ),
                        dbc.Label("Number of matches", width="auto"),
                        dbc.Col(
                            dbc.Input(
                                id="arc-num-matches",
                                type="number",
                                min=2,
                                max=10,
                                step=1,
                                value=3,
                            ),
                        ),
                        dbc.Label("Threshold", width="auto"),
                        dbc.Col(
                            dbc.Input(
                                id="arc-threshold",
                                type="number",
                                min=0.1,
                                max=1,
                                step=0.1,
                                value=0.3,
                            ),
                        ),
                        dbc.Col(dbc.Button("Map to ARC", id="map-btn"), width="auto"),
                    ],
                    className="g-2",
                ),
            ]
        ),
        style={
            "border": "1px solid silver",
            "border-radius": "0.4em",
            "padding": "1em",
        },
    ),
    style={"margin-top": "1em"},
)

output_table = dbc.Container(
    html.Div(
        dbc.Row(id="output"),
        style={"padding": "0.5em", "border": "1px solid silver", "borderRadius": "5px"},
    )
)


@callback(
    Output("upload-data-dictionary", "data"),
    Output("upload-status", "children"),
    Input("upload-btn", "n_clicks"),
    State("upload-input-file", "contents"),
    State("upload-input-file", "filename"),
    State("upload-col-responses", "value"),
    State("upload-col-description", "value"),
    prevent_initial_call=True,
)
def upload_data_dictionary(
    _,
    upload_contents,
    filename,
    col_responses,
    col_description,
):
    ok = dbc.Alert("Upload successful", color="success")
    err = lambda msg: dbc.Alert(msg, color="danger")
    if ctx.triggered_id == "upload-btn" and upload_contents is not None:
        try:
            df = read_upload_data(upload_contents, filename)
            # this is the unprocessed data dictionary, we will now convert
            # it into a standardised format
            assert df is not None
            if col_description not in df.columns:
                return {}, err("Description column not found")
            if col_responses not in df.columns:
                return {}, err("Responses column not found")
            data = read_data_dictionary(df, description_field=col_description, response_field=col_responses, response_func="redcap")
            return data.to_json(), ok

        except Exception as e:
            print(e)
            return {}, err("Upload failed due to unknown reason")
    return {}, err("Upload failed due to unknown reason")


@callback(
    Output("output", "children"),
    State("upload-data-dictionary", "data"),
    Input("map-btn", "n_clicks"),
    State("arc-version", "value"),
    State("arc-mapping-method", "value"),
    State("arc-num-matches", "value"),
    prevent_initial_call=True,
)
def invoke_map_arc(data, _, version, method, num_matches):
    if ctx.triggered_id == "map-btn":
        arc = read_arc_schema(version)
        dictionary = pd.read_json(data)
        mapped_data = map_data_dictionary_to_arc(method, dictionary, arc, num_matches)
        return (
            dash_table.DataTable(
                mapped_data.to_dict("records"),
                style_data={
                    "whiteSpace": "normal",
                    "height": "auto",
                },
                style_table={"overflowX": "auto"},
                page_size=20,
            ),
        )
    return html.Span("No data to see here")


app.layout = html.Div([navbar, upload_form, arc_form, output_table])
server = app.server
