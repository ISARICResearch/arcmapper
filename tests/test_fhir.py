from pathlib import Path

import pandas as pd

from arcmapper.fhir import FHIRMapping, merge, format_merge

MAPPING = Path(__file__).parent.parent / "arc-fhir" / "ARC_pre_1.0.0_preset_dengue.xlsx"
DRAFT_MAPPING = Path(__file__).parent / "data" / "arcmapper-mapping-file.csv"


def test_fhir_mapping():
    m = FHIRMapping(MAPPING)
    assert m.resources == [
        "Condition",
        "DiagnosticReport",
        "Encounter",
        "Immunization",
        "MedicationAdministration",
        "MedicationStatement",
        "Observation",
        "Patient",
        "Procedure",
        "Specimen",
    ]

    encounter = m.get_resource("encounter")
    assert {"arc_variable", "arc_response"} <= set(encounter.columns)


def test_merge(snapshot):
    draft_mapping = pd.read_csv(DRAFT_MAPPING)
    fhir_mapping = FHIRMapping(MAPPING)
    data = merge(draft_mapping, fhir_mapping, resources=["Patient"])
    assert format_merge(data) == snapshot
