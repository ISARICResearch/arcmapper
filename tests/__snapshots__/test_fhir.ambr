# serializer version: 1
# name: test_merge
  '''
  {{{ resource Patient
  raw_variable	raw_response	arc_variable	arc_response	raw_description	arc_description
  subjid		subjid		Participant Identification Number (PIN) specify CPMS Site code (hyphen) four to six digit number patient number e.g. Y0401-0001.	Participant Identification Number (PIN) The Participant Identification Number or PIN is a unique number used to identify the patient's (participant's) data.
  sex	1, Male	demog_sex	1, Male	Sex at Birth:	Sex at birth This refers to the physical sexual characteristics at birth.
  sex	2, Female	demog_sex	2, Female	Sex at Birth:	Sex at birth This refers to the physical sexual characteristics at birth.
  sex	3, Not specified	demog_sex	99, Not specified/Unknown	Sex at Birth:	Sex at birth This refers to the physical sexual characteristics at birth.
  age_estimateyears		demog_age		Age/Estimated age	Age This refers to the patient's age.
  age_estimateyearsu	1, Months	demog_age_units	2, Months	Age/Estimated age Unit	Age units This refers to the patient's age at presentation to the health facility.
  age_estimateyearsu	2, Years	demog_age_units	1, Years	Age/Estimated age Unit	Age units This refers to the patient's age at presentation to the health facility.
  }}}

  '''
# ---
